core/validator.go

```diff
@@ -43,74 +107,129 @@ func (v *chunkValidator) ValidateBlob(blob *BlobMessage, operatorState *Operator
-		if err != nil {
-			return err
-		}
+func (v *chunkValidator) ValidateBatch(blobs []*BlobMessage, operatorState *OperatorState, pool common.WorkerPool) error {
+	subBatchMap := make(map[EncodingParams]*SubBatch)
+	blobCommitmentList := make([]BlobCommitments, len(blobs))
 
-		if params.ChunkLength != chunkLength {
-			return errors.New("number of chunks does not match assignment")
+	for k, blob := range blobs {
+		if len(blob.Bundles) != len(blob.BlobHeader.QuorumInfos) {
+			return errors.New("number of bundles does not match number of quorums")
 		}
 
-		// Get the chunk length
-		chunks := blob.Bundles[quorumHeader.QuorumID]
-		for _, chunk := range chunks {
-			if uint(chunk.Length()) != chunkLength {
-				return ErrChunkLengthMismatch
+		// Saved for the blob length validation
+		blobCommitmentList[k] = blob.BlobHeader.BlobCommitments
+
+		// for each quorum
+		for _, quorumHeader := range blob.BlobHeader.QuorumInfos {
+			chunks, assignment, params, err := v.validateBlobQuorum(quorumHeader, blob, operatorState)
+			if err == ErrBlobQuorumSkip {
+				continue
+			} else if err != nil {
+				return err
+			} else {
+				// Check the received chunks against the commitment
+				blobIndex := 0
+				subBatch, ok := subBatchMap[*params]
+				if ok {
+					blobIndex = subBatch.NumBlobs
+				}
+
+				indices := assignment.GetIndices()
+				samples := make([]Sample, len(chunks))
+				for ind := range chunks {
+					samples[ind] = Sample{
+						Commitment:      blob.BlobHeader.BlobCommitments.Commitment,
+						Chunk:           chunks[ind],
+						AssignmentIndex: uint(indices[ind]),
+						BlobIndex:       blobIndex,
+					}
+				}
+
+				// update subBatch
+				if !ok {
+					subBatchMap[*params] = &SubBatch{
+						Samples:  samples,
+						NumBlobs: 1,
+					}
+				} else {
+					subBatch.Samples = append(subBatch.Samples, samples...)
+					subBatch.NumBlobs += 1
+				}
 			}
 		}
+	}
 
-		// Validate the chunk length
-		if chunkLength*quorumHeader.QuantizationFactor*numOperators != quorumHeader.EncodedBlobLength {
-			return ErrInvalidHeader
-		}
+	// Parallelize the universal verification for each subBatch
+	numResult := len(subBatchMap) + len(blobCommitmentList)
+	// create a channel to accept results, we don't use stop
+	out := make(chan error, numResult)
+
+	// parallelize subBatch verification
+	for params, subBatch := range subBatchMap {
+		params := params
+		subBatch := subBatch
+		pool.Submit(func() {
+			v.universalVerifyWorker(params, subBatch, out)
+		})
+	}
+
+	// parallelize length proof verification
+	for _, blobCommitments := range blobCommitmentList {
+		blobCommitments := blobCommitments
+		pool.Submit(func() {
+			v.VerifyBlobLengthWorker(blobCommitments, out)
+		})
+	}
 
-		// Check the received chunks against the commitment
-		err = v.encoder.VerifyChunks(chunks, assignment.GetIndices(), blob.BlobHeader.BlobCommitments, params)
+	for i := 0; i < numResult; i++ {
+		err := <-out
 		if err != nil {
 			return err
```

Comment:
Do you want to stop() the workerpool before return, since there is no point for remaining threads to continue?

Reply:
the workerpool is created by node thread, so it does not make a lot sense to stop them.